text/0037-guarded-arith.md

```diff
@@ -0,0 +1,82 @@
+
+```
+forbid(principal, action, resource) when {
+   principal.elevated_privileges_end_time_s * 1000 > context.current_time_ms ||
+   !principal.is_mfa
+};
+```
+
+A casual reader of this policy may conclude that all actions are forbiden when `principal.mfa` is `false`. This is wrong. The lhs (`principal.elevated_privileges_end_time_s * 1000 > context.current_time_ms`) is evaluated first and if it errors the entire policy is skipped.
+
+To avoid this class of mistakes, we want to change Cedar's validator to error on arithmetic expressions that could overflow. To pass validation, arithmetic should be guarded. For example, the above policy should be written:
+
+```
+forbid(principal, action, resource) when {
+   (principal.elevated_privileges_end_time_s < i64::MAX / 1000 &&
+    principal.elevated_privileges_end_time_s > i64::MIN / 1000 &&
+    principal.elevated_privileges_end_time_s * 1000 > context.current_time_ms) ||
+   !principal.is_mfa
+};
+```
+
+## Detailed design
+
+Every non-identity preserving arithmetic operation should be guarded by comparisons that prevent over/underflow.
+
+- We add intervals to represent the ranges of each integer.
+- Integers without explicit intervals are assumed to lie in `[i64::MIN, i64::MAX]` so almost any arithmetic operation on them will be a validation error.
+- We use interval arithmetic, which is imprecise, but because the bounds are written in the policy, the intervals can be precisely given at the point of each operation.
+
+
+## Drawbacks
+
+The policies that use arithmetic and will pass validation are ugly.
+
+We assume almost no one uses arithmetic, so we think this is an acceptable tradeoff. If there are significant uses of arithmetic, a better approach is likely desirable.
+
+## Alternatives
+
+### Unbounded Integers
+Unclear how these will affect our FFI. We can take this proposal now and switch to unbounded ints later more easily than we can switch now and reverse that decision later.
+
+### Bounds in the schema.
+It is not clear how to analyze a set with members of an EntityType that has an integer attribute with bounds while keeping analysis sound, complete and decidable.
+
+### Status quo
+See the downside in the motivation.
+
+### Less general range analysis
+Unlike https://github.com/cedar-policy/cedar/pull/56, because we're putting bounds in the policies, we could require explicit bounds for each arithmetic operation.
+
+So
+```
+(if principal.a < 9,223,372,036,854,774 && principal.a > -9,223,372,036,854,774 then principal.a * 1000 else 0) * 500 > foo
+```
+would require new bounds:
+
+```
+principal.a * 1000 + 500 < i64::MAX && (if principal.a < 9,223,372,036,854,774 && principal.a > -9,223,372,036,854,774 then principal.a * 1000 else 0) + 500 > foo
+```
+This would be easier to implement, but much more annoying to use if anyone uses chained arithmetic operations.
```

Comment:
Change "This would be easier" to "Requiring bounds for each operation would be easier ..." just so we are clear what your assessment is referring to.